### PR TITLE
[issue - #22] Abstraction to work with entity: Repository API

### PR DESCRIPTION
**Pull request by olehhaliak**

### CODE REVIEW

File: src/main/java/com/bobocode/blyznytsia/bibernate/session/Session.java [hunk 1]
```diff
@@ -19,4 +20,6 @@ public interface Session {
 
   Transaction getTransaction();
 
+  <T> Query<T> createNativeQuery(String sql, Class<T> entityType);
```
Comment (financeasist) [inline src/main/java/com/bobocode/blyznytsia/bibernate/session/Session.java:30]: java doc

File: src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    } else if (token.endsWith("IsNotNull") || token.endsWith("NotNull")) {
+      builder.isNotNull(resolveColumnName(token, "(IsNotNull|NotNull)"));
+    } else if (token.endsWith("IsNull") || token.endsWith("Null")) {
+      builder.isNotNull(resolveColumnName(token, "(IsNull|Null)"));
+    } else {
+      builder.isEqual(resolveColumnName(token, "(Is|Equals)"));
+    }
+  }
+
+
```
Comment (financeasist) [inline src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java:89]: reduntad empty line

File: src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      }
+    }
+    tokens.forEach(token -> parseSingleWhereToken(token, builder));
+  }
+
+  private void parseSingleWhereToken(String token, QueryBuilder builder) {
+    log.trace("processing token : " + token);
+    if (token == null || token.isBlank()) {
+      throw new RepositoryGenerationException("invalid token value");
+    } else if (token.equals("And")) {
```
Comment (financeasist) [inline src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java:65]: can we replace these bunch of else if with swich with pattern maching?

File: src/main/java/com/bobocode/blyznytsia/bibernate/session/SessionImpl.java [hunk 4]
```diff
@@ -183,4 +185,8 @@ private void update(Object entityIdValue, Object entity) {
 		this.persistenceContext.markForUpdate(new EntityKey(entity.getClass(), entityIdValue), entity);
 	}
 
+	@Override
```
Comment (financeasist) [inline src/main/java/com/bobocode/blyznytsia/bibernate/session/SessionImpl.java]: add   
```
/**
* {@inheritDoc}
*/
```

File: src/test/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParserTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        "SELECT * FROM person WHERE name = ?1",
+
+        "findOneByNameAndStatus",
+        "SELECT * FROM person WHERE name = ?1 AND status = ?2",
+
+        "findOneByNameOrStatus",
+        "SELECT * FROM person WHERE name = ?1 OR status = ?2",
+
+        "findAllByNameIsNull",
+        "SELECT * FROM person WHERE name IS NOT NULL",
```
Comment (financeasist) [inline src/test/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParserTest.java]: if name is `findAllByNameIsNull` why expected query is `SELECT * FROM person WHERE name IS NOT NULL`?? maybe it should be `SELECT * FROM person WHERE name IS NULL"`?

File: src/test/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParserTest.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+        "findAllByNameIsNotNull",
+        "SELECT * FROM person WHERE name IS NOT NULL",
+
+        "findAllByNameNull",
+        "SELECT * FROM person WHERE name IS NOT NULL",
```
Comment (financeasist) [inline src/test/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParserTest.java]: same as in comment above. Also if this test passed than maybe we have some mistake in the logic?

**Review by financeasist — CHANGES_REQUESTED** [review]

File: src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java (hunk 3 above)
Comment (financeasist) [reply src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java:65]: for example:

```
private void parseSingleWhereToken(String token, QueryBuilder builder) {
    if (token == null || token.isBlank()) {
      throw new RepositoryGenerationException("invalid token value");
    }

   log.trace("processing token : " + token);
    switch (token) {
      case "And" -> builder.and();
      case "Or" -> builder.or();
      case  String t && isNotNullToken(token)-> builder.isNotNull(resolveColumnName(token, "(IsNotNull|NotNull)"));
      case  String t && isNullToken(token) -> builder.isNull(resolveColumnName(token, "(IsNull|Null)"));
      default -> builder.isEqual(resolveColumnName(token, "(Is|Equals)"));
    }
  }

  private boolean isNotNullToken(String token) {
    return token.endsWith("IsNotNull") || token.endsWith("NotNull");
  }

  private boolean isNullToken(String token) {
    return token.endsWith("IsNull") || token.endsWith("Null");
  }
```

File: src/main/java/com/bobocode/blyznytsia/bibernate/session/SessionImpl.java (hunk 4 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/session/SessionImpl.java]: fixed

File: src/main/java/com/bobocode/blyznytsia/bibernate/session/Session.java (hunk 1 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/session/Session.java:30]: fixed

File: src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java (hunk 2 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java:89]: fixed

File: src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java (hunk 3 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java:65]: Unfortunately, pattern matching is a preview feature, and we've disabled it due to conflicts with other maven dependencies. I've refactored the method to be a bit more readable, hope it'll help

File: src/test/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParserTest.java (hunk 5 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParserTest.java]: Thanks for pointing that out! It was a typo, fixed.

File: src/test/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParserTest.java (hunk 6 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParserTest.java]: fixed

**Review by financeasist — APPROVED** [review]

**Review by MykolaHarasym — APPROVED** [review]

**Review by mvoronov — APPROVED** [review]

**Pull request merged**